tests/returntypes/test_gpu_torch_new.py

```diff
@@ -0,0 +1,226 @@
+        y = torch.tensor(0.2, requires_grad=True, device=torch.device("cpu"))
+
+        with qml.tape.QuantumTape() as tape1:
+            qml.RX(x, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        res1 = dev.execute(tape1)
+        assert "cuda" in dev._torch_device
+        assert res1.is_cuda
+
+        with qml.tape.QuantumTape() as tape2:
+            qml.RY(y, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        res2 = dev.execute(tape2)
+        assert dev._torch_device == "cpu"
+        assert not res2.is_cuda
+
+    def test_integration(self):
+        """Test cuda supported when device created in qnode creation."""
+
+        dev = qml.device("default.qubit", wires=1)
+
+        x = torch.tensor(0.1, requires_grad=True, device=torch.device("cuda"))
+        y = torch.tensor(0.2, requires_grad=True)
+
+        @qml.qnode(dev, interface="torch", diff_method="backprop")
+        def circ(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        res = circ(x, y)
+        assert res.is_cuda
+
+        assert "cuda" in circ.device._torch_device
+        res.backward()
+        assert x.grad.is_cuda
+
+    @pytest.mark.parametrize("init_device, par_device", [("cpu", "cuda"), ("cuda", "cpu")])
+    def test_different_devices_creation_and_parameters_warn(self, init_device, par_device):
+        """Test that a warning is raised if the Torch device specified on
+        PennyLane device creation differs from the Torch device of gate
+        parameters.
+        """
+        dev = qml.device("default.qubit.torch", wires=1, torch_device=init_device)
+
+        p = torch.tensor(0.543, dtype=torch.float64, device=par_device)
+
+        @qml.qnode(dev, interface="torch")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliY(0))
+
+        with pytest.warns(
+            UserWarning,
+            match=f"Torch device {init_device} specified upon PennyLane device creation does not match",
+        ):
+            circuit(p)
+
+    def test_amplitude_embedding(self):
+        """Test that the padding capability of amplitude embedding works with
+        GPUs."""
+        n_wires = 2
+        dev = qml.device("default.qubit", wires=n_wires)
+
+        @qml.qnode(dev, interface="torch")
+        def circuit_cuda(inputs):
+            qml.AmplitudeEmbedding(inputs, wires=range(n_wires), pad_with=0, normalize=True)
+            return qml.expval(qml.PauliZ(0))
+
+        inputs = torch.rand(n_wires)  # embedding shorter than 2 ** n_wires
+        res1 = circuit_cuda(inputs)
+        assert not res1.is_cuda
+
+        inputs = inputs.to(torch.device("cuda"))  # move to GPU
+        res2 = circuit_cuda(inputs)
+
+        assert res2.is_cuda
+
+
+# TODO: Add those tests after support for TorchLayers
+@pytest.mark.xfail(reason="Add Torch layer to the new return type system.")
```

Comment:
Is this currently on our radar?

Reply:
After Pytorch is merged yes it is a TODO same for TF.